### PR TITLE
Mac tracker refactoring

### PR DESCRIPTION
**Pull request by royru**
Hi Chris, hi André

Here's the PR I mentioned in the discussion last week. It features some refactoring work to better encapsulate the trackers. #208 

Sorry in advance for the huge number of lines changed. I hope PRs will be smaller in the future ;) 

- **Tracker Renaming**: All tracker were renamed (if necessary) to match the naming in the Windows version. You'll see the new names [here](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-44ff70bd9144de94ee552e7e42df308a) or when opening the workspace in xcode.

- **Tracker Classes Renaming**: Every tracker comprises at least 3 files: `<name>Tracker.swift` the main class/file of the tracker (Daemon.cs in the windows version), `<name>Settings.swift` an enum with the db table name and other constants (Settings.cs in the windows version), `<name>Queries.swift` with all the sql queries. `<name>` is the tracker name (e.g. UserEfficiency, UserInput, etc.) which I had to prepend since swift does not have namespacing unfortunately...

- **shared/DataObjectController.swift**: the class/file with the respective name was deleted.
- **shared/SQLController.swift**: the class/file with the respective name was deleted. It comprised methods used for csv-export (no longer needed) and the utility methods `getStartHour` and `getEndHour`. Both were moved to the [Date extension](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-aa6cafb9761442ed4cf8a55d789ddd1d).

- **DatabaseController.swift**: the `DatabaseController` class was formerly part of the DataObjectController.swift file and has now its own file.

- **export csv**: deleted functionality, as discussed (see [emotion pop-up](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-bea1ba2d3299e96265b2a4fb127a2adb), its [controller](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-d63424be50e906de142b4fffbf3682d3) and all `fetch<...>Since` methods which were previously in [SQLController.swift](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-248cdfe45a7bbba4f4ff30a12e697c66).

- **NSNotifications**: There were some instances where a NSNotification was posted without an observer listening (e.g. in the [MouseActionController](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-e57e755b5804a9a6b13a8234959ad804). I removed those lines.

- **NSUserNotifications**: the responses to user notifications (the UserEfficiencyTracker and EmotionTracker show pop-ups upon clicking on the notification) were previously routed in `AppDelegate` with a `switch` statement. Consequentially, new trackers would need to modify AppDelegate if they wanted to use NSUserNotifications. To solve this problem I created a protocol `TrackerUserNotificationHandling` with a method `handleUserNotification()`. Trackers can implement this protocol and when the `notification.identifier` is equal to the tracker name the `TrackerManager` forwards the notification response automatically to the tracker without any further code changes necessary. See [TrackerManager.handleTrackerUserNotifications()](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-4c8a817f959c297b0d5229e2af71fe22) and the [protocol](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-225ebb0b15087d65f90269f17bca90ac).

- **UserEfficiencyTracker popup handling**: The popup to track user efficiency was previously managed in the `AppDelegate` class. I moved the functionality into the UserEfficiencyTracker [SummaryViewController](https://github.com/sealuzh/PersonalAnalytics/pull/260/files#diff-a15a0948a59cbe96334d3fd37c272125).

### CODE REVIEW

**Comment by csatterfield** [comment]
Thanks Roy! Sorry in advance it will probably take me quite a while before I have time to get through all of this.

**Comment by royru** [comment]
Sure. No stress:) Let me know if I can help you walking you through the code or in any other form! 

Sorry again for this rather messy PR

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/UserInputTracker/Visualizations/ActivityVisualization.swift [hunk 1]
```diff
@@ -18,11 +18,8 @@ class ActivityVisualization: IVisualization{
     //TODO
     var title: String
     var _type = [VisConstants.Day]
-    
-    let sql: InputSQLController
-    
+        
     required init() throws {
```
Comment (csatterfield) [inline src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/UserInputTracker/Visualizations/ActivityVisualization.swift]: not super important, but I dont think that the `throws` is necessary for any of the visualizations anymore? We should remove this

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/WindowsActivityTracker/WindowsActivityTracker.swift [hunk 2]
```diff
@@ -38,7 +34,7 @@ class ActiveApplicationTracker: ITracker{
   
     init(){
         isIdle = false
-        name = "ActiveApplication"
+        name = "WindoesActivityTracker"
```
Comment (csatterfield) [inline src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/WindowsActivityTracker/WindowsActivityTracker.swift]: typo here

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TrackerManager.swift [hunk 3]
```diff
@@ -75,6 +69,32 @@ class TrackerManager {
         return trackers[tracker]!
```
Comment (csatterfield) [inline src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TrackerManager.swift]: This is a little bit scary, and I don't remember why we have this? Can we remove it or think of a better way to access the trackers :)

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TaskProductivity Tracker/UserEfficiencyTracker.swift [hunk 4]
```diff
@@ -1,18 +1,14 @@
 //
-//  TaskProductivityTracker.swift
+//  UserEfficiencyTracker.swift
```
Comment (csatterfield) [inline src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TaskProductivity Tracker/UserEfficiencyTracker.swift]: Please also rename the folder containing the UserEfficiencyTracker

**Comment by csatterfield** [comment]
Sorry for the long delay @royru . Great work! Just a few small comments again.

File: src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift [hunk 5]
```diff
@@ -290,11 +234,9 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSUserNotificationCenterDele
         center.delegate = self
         
         
-        addNotificationListeners()
         // Start local server so chrome extension can send data to it
```
[the remark this answers is not in the document]
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift]: Removed the part about "chrome extension".

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TaskProductivity Tracker/SummaryViewController.swift [hunk 6]
```diff
@@ -8,29 +8,44 @@
...
+    func closeSummaryPopup(){
+        if(popover.isShown){
+            popover.performClose(nil)
+        }
     }
     
+    func showSummaryPopup() {
+        popover.contentViewController = self
+        popover.behavior = .transient
+        // TODO: Not nice to use the delegate here, better options?
```
[the remark this answers is not in the document]
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TaskProductivity Tracker/SummaryViewController.swift]: Agreed. I would go for a regular window which opens up after the user acknowledged the notification. The emotion tracker uses a regular window as well for the small emotion survey and it feels natural. By sending a notification up front we are not interrupting/confusing the user with a window popping up out-of-the-blue as he/she expects something to happen after acknowledging the notification.

 If I remember correctly, the Windows version shows a pop up in the bottom right corner (but without sending a notification first).

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/UserInputTracker/Visualizations/ActivityVisualization.swift (hunk 1 above)
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/UserInputTracker/Visualizations/ActivityVisualization.swift]: You are right! removed it.

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/WindowsActivityTracker/WindowsActivityTracker.swift (hunk 2 above)
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/WindowsActivityTracker/WindowsActivityTracker.swift]: thx,  fixed

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TrackerManager.swift (hunk 3 above)
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TrackerManager.swift]: True. At first I wanted to get rid of `getTracker` (its' only used once in the emotion tracker).  Unfortunately, I did not come up with an elegant way to accomplish what was required.

I settled with returning the optional value (see latest commits). With that modification, a tracker can still run into problems if the wrong tracker name is passed as an argument. But at least the tracker can handle that case itself.

File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TaskProductivity Tracker/UserEfficiencyTracker.swift (hunk 4 above)
Comment (royru) [reply src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/TaskProductivity Tracker/UserEfficiencyTracker.swift]: I will check and rename all folders (if necessary)  in a separate PR soon. I have some other code changes open and don't want to run into a bunch of merge conflicts ;)

**Comment by royru** [comment]
Thanks a lot for the feedback! This time I had a few busy weeks :)  

Let me know if my option regarding the "popover" works for you! I would then make the changes in a separte PR.

**Comment by csatterfield** [comment]
Thank Roy! I agree with your suggestion to use a regular window for the popup. This is definitely low priority though, don't spend too much time on this 😄

**Pull request merged**